certbot/ocsp.py

```diff
@@ -65,33 +92,117 @@ def ocsp_revoked(self, cert_path, chain_path):
+    if response.status_code != 200:
+        logger.info("OCSP check failed for %s (are we offline?)", cert_path)
+        return False
+    response_ocsp = ocsp.load_der_ocsp_response(response.content)
+
+    # Check OCSP response validity
+    if response_ocsp.response_status != ocsp.OCSPResponseStatus.SUCCESSFUL:
+        logger.error("Invalid OCSP response status for %s: %s",
+                     cert_path, response_ocsp.response_status)
+        return False
+
+    # Check OCSP signature
+    try:
+        _check_ocsp_response(response_ocsp, request, issuer)
+    except UnsupportedAlgorithm as e:
+        logger.error(str(e))
+    except errors.Error as e:
+        logger.error(str(e))
+    except InvalidSignature:
+        logger.error('Invalid signature on OCSP response for %s', cert_path)
+    except AssertionError as error:
+        logger.error('Invalid OCSP response for %s: %s.', cert_path, str(error))
+    else:
+        # Check OCSP certificate status
+        logger.debug("OCSP certificate status for %s is: %s",
+                     cert_path, response_ocsp.certificate_status)
+        return response_ocsp.certificate_status == ocsp.OCSPCertStatus.REVOKED
+
+    return False
+
+
+def _check_ocsp_response(response_ocsp, request_ocsp, issuer_cert):
+    """Verify that the OCSP is valid for serveral criterias"""
+    # Assert OCSP response corresponds to the certificate we are talking about
+    if response_ocsp.serial_number != request_ocsp.serial_number:
+        raise AssertionError('the certificate in response does not correspond '
+                             'to the certificate in request')
+
+    # Assert signature is valid
+    _check_ocsp_response_signature(response_ocsp, issuer_cert)
+
+    # Assert issuer in response is the expected one
+    if (not isinstance(response_ocsp.hash_algorithm, type(request_ocsp.hash_algorithm))
+            or response_ocsp.issuer_key_hash != request_ocsp.issuer_key_hash
+            or response_ocsp.issuer_name_hash != request_ocsp.issuer_name_hash):
+        raise AssertionError('the issuer does not correspond to issuer of the certificate.')
+
+    # Assert nextUpdate is in the future, and that thisUpdate is not too old
+    if response_ocsp.next_update:
+        if response_ocsp.next_update < datetime.now() - timedelta(minutes=5):
+            raise AssertionError('next update is in the past.')
+        interval = response_ocsp.next_update - response_ocsp.this_update
+        if datetime.now() - response_ocsp.this_update > interval + timedelta(minutes=5):
+            raise AssertionError('this update is too old.')
+
+
+def _check_ocsp_response_signature(response_ocsp, issuer_cert):
+    """Verify an OCSP response signature against certificate issuer"""
+    # Following line may raise UnsupportedAlgorithm
+    chosen_hash = response_ocsp.signature_hash_algorithm
+    crypto_util.verify_signed_payload(issuer_cert.public_key, response_ocsp.signature,
```

Comment:
```suggestion
    crypto_util.verify_signed_payload(issuer_cert.public_key(), response_ocsp.signature,
```

Without this change, I'm getting:
```
Unsupported public key type
```
when doing manual testing.

Reply:
Indeed. Corrected.